File: Programmers/Lv2/Lv2_타겟넘버.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int answer = 0;

int oper(int i, int num, int sum) {
	if (i == 0) {				// 더하기
		sum += num;
	}
	else if (i == 1) {			// 빼기
		sum -= num;
	}
	return sum;
}

void go(vector<int> numbers, int target, int index,int result) {
	if (result == target && index == numbers.size()) {			// 완료조건
		answer++;
		return;
	}
	if (index > numbers.size() - 1)								// 종료조건
		return;
	for (int i = 0; i < 2; i++) 
		go(numbers, target, index + 1, oper(i, numbers[index], result));
}

int solution30(vector<int> numbers, int target) {
	go(numbers,target,0,0);
	
	return answer;
}

File: Programmers/Lv2/Lv2_타겟넘버.py
answer = 0

def oper(i,num,sum):
    if i==0:
        sum += num
    elif i==1:
        sum -= num
    return sum

def go(numbers, target, index, result):
    global answer
    if result == target and index == len(numbers):
        answer += 1
        return
    if index > len(numbers)-1:
        return
    go(numbers,target, index+1, oper(0,numbers[index], result))
    go(numbers,target, index+1, oper(1,numbers[index], result))

def solution(numbers, target):
    go(numbers,target,0,0)
    return answer

print(solution([1, 1, 1, 1, 1], 3))